### PR TITLE
OHRM5X-1198: Develop admin - export language package XLIFF

### PR DESCRIPTION
**Pull request by ManulMax**
Remove 'translated' column in 'ohrm_i18n_translate' table for the new 5.2 version.

### CODE REVIEW

**Review by Super-Chama — APPROVED** [review]

File: src/plugins/orangehrmAdminPlugin/Dao/LocalizationDao.php [hunk 1]
```diff
@@ -90,6 +90,7 @@ public function getLanguageById(int $languageId): ?I18NLanguage
      */
     public function saveI18NLanguage(I18NLanguage $i18NLanguage): I18NLanguage
     {
+        $i18NLanguage->getModifiedAt();
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Dao/LocalizationDao.php:93]:
```suggestion
```

File: src/plugins/orangehrmAdminPlugin/Dao/LocalizationDao.php [hunk 2]
```diff
@@ -129,12 +130,10 @@ private function getTranslationsQueryBuilderWrapper(
 
         if (!is_null($i18NTargetLangStringSearchFilterParams->getOnlyTranslated())) {
             if ($i18NTargetLangStringSearchFilterParams->getOnlyTranslated() === true) {
-                $q->andWhere($q->expr()->isNotNull('translation.value'));
-                $q->andWhere($q->expr()->isNotNull('translation.translated'));
                 $q->andWhere($q->expr()->notIn('translation.value', ':translated'));
                 $q->setParameter('translated', '');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Dao/LocalizationDao.php:134]:
```suggestion
                $q->andWhere($q->expr()->isNotNull('translation.value'));
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @return LocalizationService
+     */
+    public function getLocalizationService(): LocalizationService
+    {
+        if (!$this->localizationService instanceof LocalizationService) {
+            $this->localizationService = new LocalizationService();
+        }
+        return $this->localizationService;
+    }
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php:48]: use the service from service container

File: src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @param Request $request
+     * @return Response
+     */
+    public function handle(Request $request): Response
+    {
+        $languageId = $request->attributes->get('languageId');
+        $response = $this->getResponse();
+
+        if ($languageId) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php:59]:
```suggestion
        if ($request->attributes->has('languageId')) {
```

File: src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        return $this->localizationService;
+    }
+
+    /**
+     * @param Request $request
+     * @return Response
+     */
+    public function handle(Request $request): Response
+    {
+        $languageId = $request->attributes->get('languageId');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php:56]:
```suggestion
```
move this inside to if block

File: src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        $response = $this->getResponse();
+
+        if ($languageId) {
+            $language = $this->getLocalizationService()->getLocalizationDao()
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php:60]:
```php
            $languageId = $request->attributes->getInt('languageId');
            $language = $this->getLocalizationService()->getLocalizationDao()
```

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 7]
```diff
@@ -192,4 +199,78 @@ public function generateLangStringLanguageKey(int $languageId, int $langStringId
...
             $langStringId . '_';
     }
+
+    /**
+     * @param string $languageId
+     * @param string $langCode
+     * @return false|string
+     * @throws \DOMException
+     */
+    public function exportLanguagePackage(string $languageId, string $langCode)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:209]: define return type

File: src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                ->getLanguageById($languageId);
+
+            if (!($language instanceof I18NLanguage)
+                || !$language->isEnabled()
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/File/LanguagePackage.php:64]: check  added packs also

File: src/plugins/orangehrmAdminPlugin/Dao/LocalizationDao.php [hunk 9]
```diff
@@ -111,6 +111,25 @@ public function getNormalizedTranslations(
...
+        $q = $this->getTranslationsQueryBuilderWrapper($i18NTargetLangStringSearchFilterParams)->getQueryBuilder();
+        $q->select(
+            'langString.id AS langStringId',
+            'langString.unitId AS unitId',
+            'langString.value AS source',
+            'langString.note AS note',
+            'langString.version AS version',
+            'translation.value AS target',
+        );
+        return array_values($q->getQuery()->execute());
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Dao/LocalizationDao.php:130]:
```suggestion
        return $q->getQuery()->execute();
```
:thinking:

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 10]
```diff
@@ -192,4 +199,78 @@ public function generateLangStringLanguageKey(int $languageId, int $langStringId
...
+    }
+
+    /**
+     * @param $i18nGroups
+     * @param $langCode
+     * @param $languageId
+     * @return DOMDocument
+     * @throws \DOMException
+     */
+    public function getXliffXmlSources($i18nGroups, $langCode, $languageId): DOMDocument
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:225]:
```suggestion
    private function getXliffXmlSources($i18nGroups, $langCode, $languageId): DOMDocument
```
define arg types

File: src/client/src/orangehrmAdminPlugin/pages/languagePackage/languagePackage.vue [hunk 11]
```diff
@@ -192,6 +193,10 @@ export default {
     onClickTranslate(item) {
       navigate('/admin/languageCustomization/{id}', {id: item.id});
     },
+    onClickExport(item) {
+      const url = `${window.appGlobal.baseUrl}/admin/viewLanguagePackage/languageId/${item.id}`;
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/languagePackage/languagePackage.vue:197]: *

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Review by Super-Chama — CHANGES_REQUESTED** [review]
The file
`src/client/src/orangehrmAdminPlugin/pages/languagePackage/languagePackage.vue`
should be renamed in pascal case as per our project convention. `src/client/src/orangehrmAdminPlugin/pages/languagePackage/LanguagePackage.vue`

Please update the file name and all imports of that file.

File: src/client/src/orangehrmAdminPlugin/pages/languagePackage/languagePackage.vue (hunk 11 above)
Comment (Super-Chama) [reply src/client/src/orangehrmAdminPlugin/pages/languagePackage/languagePackage.vue:197]: this is okay. we have same usage everywhere else with downloads.

File: installer/Migration/V5_2_0/LangStringHelper.php [hunk 12]
```diff
@@ -24,6 +24,7 @@
 use OrangeHRM\Installer\Util\V1\Dto\LangString;
 use OrangeHRM\Installer\Util\V1\LanguageHelper;
 
+//TODO:: need to refactor this class
```
Comment (Sharukap) [inline installer/Migration/V5_2_0/LangStringHelper.php:27]: @RajithaKumara I'll handle this in my next PR. Just need to have a common class neh?

**Review by Sharukap — APPROVED** [review]

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 13]
```diff
@@ -192,4 +200,75 @@ public function generateLangStringLanguageKey(int $languageId, int $langStringId
         return $languageId . '_' .
             $langStringId . '_';
     }
+
+    /**
+     * @param I18NLanguage $language
+     * @return String
+     */
+    public function exportLanguagePackage(I18NLanguage $language): String
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:208]:
```suggestion
    public function exportLanguagePackage(I18NLanguage $language): string
```

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 14]
```diff
@@ -192,4 +200,75 @@ public function generateLangStringLanguageKey(int $languageId, int $langStringId
         return $languageId . '_' .
             $langStringId . '_';
     }
+
+    /**
+     * @param I18NLanguage $language
+     * @return String
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:206]:
```suggestion
     * @return string
```

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 15]
```diff
@@ -192,4 +200,75 @@ public function generateLangStringLanguageKey(int $languageId, int $langStringId
...
+    {
+        $i18NGroupSearchFilterParams = new I18NGroupSearchFilterParams();
+        $i18nGroups = $this->getLocalizationDao()->searchGroups($i18NGroupSearchFilterParams);
+        $i18nSources = $this->getXliffXmlSources($i18nGroups, $language);
+
+        return $i18nSources->saveXML();
+    }
+
+    /**
+     * @param array $i18nGroups
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:218]:
```suggestion
     * @param I18NGroup[] $i18nGroups
```

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**